src/Features/LanguageServer/Protocol/Handler/Diagnostics/DocumentEditAndContinueDiagnosticSourceProvider.cs

```diff
@@ -0,0 +1,29 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Immutable;
+using System.Composition;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.EditAndContinue;
+using Microsoft.CodeAnalysis.Host.Mef;
+
+namespace Microsoft.CodeAnalysis.LanguageServer.Handler.Diagnostics;
+
+[Export(typeof(IDiagnosticSourceProvider)), Shared]
+[method: ImportingConstructor]
+[method: Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+internal sealed class DocumentEditAndContinueDiagnosticSourceProvider()
+    : AbstractDocumentDiagnosticSourceProvider(PullDiagnosticCategories.EditAndContinue)
```

Comment:
i would prefer this not use subclassign.   Imagine it more like a feature written by an external team.  We try to not use subclassing for extensiblity.

Reply:
What defines "extensibility"? Why such inheritance is discouraged? I see similar examples, e.g. `AbstractCodeCleanupService`, `AbstractExecuteWorkspaceCommandHandler`, `AbstractLspCompletionResultCreationService`, etc.

Reply:
> What defines "extensibility"? 

Two disparate feature areas interacting through some established extension point.  

> AbstractCodeCleanupService

This is to allow VB/C# to export this type.  We use inheritance to allow the same feature area to customize across our specific roslyn layers.  in this case we layer as:

compiler
workspace
features
editorfeatures
visualstudio

And at each of those layers we have:  core, c#, vb

So a specific feature needing to customize a bit up/down that layering stack, or side-to-side across core/c#/vb is ok (and we try to be very controlled even there).

> AbstractExecuteWorkspaceCommandHandler

I don't know what this type is.  But it seems to only have a single subclass.  I would remove it.

> AbstractLspCompletionResultCreationService

This falls into the layering discussion above.  It allows different layers *of the same feature* to change behavior.  It's ntoa  general *open ended* extension mechanism.

--

What we have here is different.  We have the core diagnostics system.  Andwe have an extension system fro arbitrary, and unbounded, features to come play in that space.  We do not generally use inheritance for that.